ios emulation-based analysis (#1553)
OWASP/owasp-mstg: Document/0x06c-Reverse-Engineering-and-Tampering.md

 21324 ms     | -[NSURLRequest initWithURL:0x106388b00 cachePolicy:0x0 timeoutInterval:0x106388b80
```

#### Emulation-based Analysis

Emulation is imitation of a certain computer platform or program on another platform or program. The software or hardware performing this imitation is called an _emulator_. Emulator provides a much cheaper alternatives to an actual device, where a user can manipulate it without worrying about damaging the device. As discussed in "[Emulation-based analysis](0x05c-reverse-engineering-and-tampering#emulation-based-analysis "Emulation-based analysis")" section, there are multiple emulators available for Android, but unlike Android, for iOS practically there are practically no viable emulators available.

##### iOS Simulator

> Apple provides a simulator app within Xcode which provides an user interface for iPhone, iPad or Apple Watch. As per Apple documentation, "simulator allows you to rapidly prototype and test builds of your app during the development process." Although the iOS simulator provides a similar look and feel as a real iOS device and also allows to run a debug application, but actually it is not an emulator. Difference between a simulator and an emulator often causes confusion and leads to use of the two terms interchangeably, but in reality they are different, specially for our current use case.

Reviewer: I'd condensate this and move the last part to the next paragraph:

```suggestion
Apple provides a simulator app within Xcode which provides a _real iOS device looking_ user interface for iPhone, iPad or Apple Watch. It allows you to rapidly prototype and test debug builds of your applications during the development process, but actually **it is not an emulator**.
```

> An emulator mimics both the software and hardware environment of a targeted platform. On the other hand, a simulator only mimics the software environment. QEMU based emulators for Android take into consideration the RAM, CPU, battery performance etc (hardware components) while running an application, but in an iOS simulator these hardware component behaviour is not taken into consideration at all. In simple words, emulator is a much closer imitation of the targeted platform, while simulator mimics only a part of it. The iOS simulator even lacks the implementation of the iOS kernel, as a result if an application is using syscalls it cannot be executed in this simulator.

Reviewer: I'd also put this in 4c. It also gives a nice general overview.

I've added part of the previous paragraph here.

```suggestion
The difference between a simulator and an emulator often causes confusion and leads to use of the two terms interchangeably, but in reality they are different, specially for the iOS use case. An emulator mimics both the software and hardware environment of a targeted platform. On the other hand, a simulator only mimics the software environment.

QEMU based emulators for Android take into consideration the RAM, CPU, battery performance etc (hardware components) while running an application, but in an iOS simulator this hardware component behaviour is not taken into consideration at all. The iOS simulator even lacks the implementation of the iOS kernel, as a result if an application is using syscalls it cannot be executed in this simulator.

In simple words, an emulator is a much closer imitation of the targeted platform, while a simulator mimics only a part of it. 
```

> iOS simulator only runs on an x86 platform, and as a consequence applications having code compiled in x86 instruction set only can be executed on it. While developing and debugging an application, Xcode toolchain do generate x86 code, but for a release build the code is compiled into ARM instruction set only, i.e no x86 code is generated. Due to the absence of x86 code in the release build, an application downloaded from Apple App Store cannot be executed on an iOS simulator and thus cannot be used for any kind of application analysis.

Reviewer: I'd leave this here but condensate it in shorter sentences, what do you think of this?

```suggestion
While developing and debugging an application, the Xcode toolchain generates x86 code, which can be executed in the iOS simulator. However, for a release build, only ARM code is generated (incompatible with the iOS simulator). That's why applications downloaded from the Apple App Store cannot be used for any kind of application analysis on the iOS simulator.
```

> In summary, iOS simulator is not an emulator and an application from Apple App Store cannot be executed as it does not have the x86 compiled code.

Reviewer: I'd remove this completely as, after the other changes the section will be very straightforward and therefore doesn't need a summary ;)
```suggestion
```

##### Corellium

> Corellium is a commercial tool which offers virtual iOS devices - a virtual iPhone running actual iOS firmware. The product is very expensive (IDA-Pro license will feel very cheap) and there are no trial licenses available. This product is discussed here as it is the only publicly available iOS emulator ever.

Reviewer 2: See main comment

> Corellium is a commercial tool which offers virtual iOS devices - a virtual iPhone running actual iOS firmware. The product is very expensive (IDA-Pro license will feel very cheap) and there are no trial licenses available. This product is discussed here as it is the only publicly available iOS emulator ever.

Reviewer: I've integrated a part of the next paragraph here and made clear that it is an emulator (at the beginning already, so no more explanation would be required).

The basic message for this and next paragraph: "it's an actual emulator that can be used to analyze release apps but we won't go into detail as it is commercial".

```suggestion
Corellium is a commercial tool which offers virtual iOS devices running actual iOS firmware, being the only publicly available iOS emulator ever. Since it is a proprietary product, not much information is available about the implementation. Corellium has an elevated price and there are no trial or community licenses available, therefore we won't go into much detail regarding its use.
```

Author: can we use other word than "elevated price". I have never seen elevated used in such context? do you agree?

Reviewer: I agree and adapted the suggestion. What do you think?

> Since it is a proprietary product, not much information is available about the implementation, but it is safe to assume that some kind of emulation is used behind the scenes. For an end user it is an emulator, with which multiple instances of a device can be launched, and the ability to take a snapshot of the device state and restore it, if needed. The product is quite straightforward to use and the virtual device can be accessed as a local device with some VPN configurations, thus making it available for application development as well.

Reviewer: ```suggestion
Corellium allows you to launch multiple instances of a device (jailbroken or not) which are accessible as local devices (with a simple VPN configuration). It has the ability to take and restore snapshots of the device state, and also offers a convenient web-based shell to the device. Finally and most importantly, due to its "emulator" nature, you can execute applications downloaded from the Apple App Store, enabling any kind of application analysis as you know it from real iOS (jailbroken) devices.
```

Author: nice!

#### Emulation-based Analysis

##### iOS Simulator

> Apple provides a simulator app within Xcode which provides a _real iOS device looking_ user interface for iPhone, iPad or Apple Watch. It allows you to rapidly prototype and test debug builds of your applications during the development process, but actually **it is not an emulator**.

Reviewer: After doing what's mentioned below I'd end this paragraph with a reference to 0x4c section. Something like: `See section "[Emulation-based Dynamic Analysis](...)" for more details.`

> The difference between a simulator and an emulator often causes confusion and leads to use of the two terms interchangeably, but in reality they are different, specially for the iOS use case. An emulator mimics both the software and hardware environment of a targeted platform. On the other hand, a simulator only mimics the software environment.

Reviewer: I think I'd actually take from here 445 until 449 and put this explanation in 0x4c as well. It's very nice to have it in one place talking about both platforms and to keep this section short. That would leave us here with only 2 nice paragraphs going straight to the point ;)
